`gh status`: show status even if a comment 404s

rwe opened the pull request:
Previously, if `gh status` encountered a mention in a comment whose URL returned a 404, the entire status would fail. However, a 404'd comment can be benign, for example being part of a repo that is no longer visible. In this case, other comments and status items should continue to load.

Fixes #7872

rwe committed:
gh status: show status even if a comment 404s

Previously, if `gh status` encountered a mention in a comment whose URL
returned a 404, the entire status would fail. However, a 404'd comment
can be benign, for example being part of a repo that is no longer
visible. In this case, other comments should continue to load.

Fixes cli/cli#7872

andyfeller commented:
@rwe : Thanks for opening this pull request and contributing back! ✨ Thank you for your patience while I catch up on these changes 🙇

andyfeller commented on pkg/cmd/status/status.go:
```diff
@@ -286,23 +286,35 @@ func (s *StatusGetter) LoadNotifications() error {
 						return nil
 					}
-					if actual, err := s.ActualMention(n.Subject.LatestCommentURL); actual != "" && err == nil {
+					var preview string
+					if actual, err := s.ActualMention(n.Subject.LatestCommentURL); err == nil {
+						preview = actual
+					} else {
+						var httpErr api.HTTPError
+						if !errors.As(err, &httpErr) {
+							abortFetching()
+							return fmt.Errorf("could not fetch comment: %w", err)
+						}
+						if httpErr.StatusCode == 403 {
+							s.addAuthError(httpErr.Message, factory.SSOURL())
+							return nil
+						} else if httpErr.StatusCode == 404 {
+							preview = "(could not find comment)"
+						} else {
+							abortFetching()
+							return fmt.Errorf("could not fetch comment: %w", err)
+						}
+					}
+
+					if preview != "" {
```
**Would you find this more readable?** 🙇

I agree with you that a `404` comment means the user cannot fetch it shouldn't stop the `gh status` output. 👍  Aside from a minor change in the wording &mdash; _(keeping the verbiage consistent and it isn't like we can't find it but can't fetch it)_ &mdash; I think this is more idiomatic handling the error immediately while also allowing us to handle http and non-http errors without duplicating the outcomes.

```suggestion
					actual, err := s.ActualMention(n.Subject.LatestCommentURL)

					if err != nil {
						var httpErr api.HTTPError
						httpStatusCode := -1

						if errors.As(err, &httpErr) {
							httpStatusCode = httpErr.StatusCode
						}

						switch httpStatusCode {
						case 403:
							s.addAuthError(httpErr.Message, factory.SSOURL())
							return nil
						case 404:
							actual = "(could not fetch comment)"
						default:
							abortFetching()
							return fmt.Errorf("could not fetch comment: %w", err)
						}
					}

					if actual != "" {
```

rwe replied:
Sure — your suggestion looks pretty equivalent to me. I also didn't like the duplicated `fmt.Errorf` in mine, but was hesitant about a mutated + dummy status code var vs single assignment, and then decided not to spend more time on it. I'll add your changes.

andyfeller commented on pkg/cmd/status/status.go:
```diff
@@ -286,23 +286,35 @@ func (s *StatusGetter) LoadNotifications() error {
 							Repository: n.Repository.FullName,
 							Identifier: fmt.Sprintf("%s#%s", n.Repository.FullName, split[len(split)-1]),
-							preview:    actual,
+							preview:    preview,
```
```suggestion
							preview:    actual,
```

andyfeller commented on pkg/cmd/status/status_test.go:
```diff
@@ -295,6 +295,63 @@ func TestStatusRun(t *testing.T) {
+			httpStubs: func(reg *httpmock.Registry) {
+				reg.Register(
+					httpmock.GraphQL("UserCurrent"),
+					httpmock.StringResponse(`{"data": {"viewer": {"login": "jillvalentine"}}}`))
```
I enjoy Resident Evil, too 🧟

rwe replied:
🤷 Didn't catch the reference, actually! Just based on the surrounding stubs.

rwe committed:
gh status: handle load errors via switch

Incorporating review changes from https://github.com/cli/cli/pull/7873/files#r1303044770

Co-authored-by: Andy Feller <andyfeller@github.com>

andyfeller commented:
@rwe : thank you for your patience!  going to run the tests now but there are some changes on `trunk` that will need to be updated here before status checks can be cleared.

samcoe commented on pkg/cmd/status/status.go:
```diff
@@ -286,7 +286,29 @@ func (s *StatusGetter) LoadNotifications() error {
+							s.addAuthError(httpErr.Message, factory.SSOURL())
+							return nil
+						case 404:
+							actual = "(could not fetch comment)"
```
Do we actually want to display anything to the user in this case? They have no actions they can take to view the comment so this message does not provide a lot of value in my opinion. I would rather see us just skip to the next one and not display anything.

rwe replied:
I also waffled on that. Personally, I prefer seeing that a notification exists even if I can't see the text. The comment URL still shows up, which gives some context, e.g. indicating there's a comment a former client meant to be seen. (I suspect that they only exist for comments made before access was restricted, not new comments tagged privately…but I don't know). In that case, the action might be to reach out to the client and let them know they might need to fill you in directly.

🤷  But I also don't feel too strongly about it.

samcoe replied:
I think the more common case of this occurring is that a comment was purposefully deleted for one reason or another and in that case showing nothing seems like the correct behavior. 

For now lets go with not displaying anything and we can change the behavior in the future if we hear from users that they would rather have the other behavior.

samcoe reviewed (changes requested):
@rwe This looks almost ready to go. I left one comment/question that I would like to see addressed before it ships.

rwe commented:
P.S. @samcoe @andyfeller do you prefer rebase or merge for trunk sync? (No conflicts).

rwe committed:
gh status: omit notifications with missing comments

Versus showing the notification with a "(could not fetch comment)"
placeholder. The reason to omit the notification entirely is that there
is little or no action a user can take to fix it, and the common case
appears to be notifications for deleted comments.

rwe committed:
Merge branch 'trunk' into fix-7872

andyfeller commented:
> P.S. @samcoe @andyfeller do you prefer rebase or merge for trunk sync? (No conflicts).

Bit of a personal choice as rebase will rewrite the history of your branch whereas merging is what I generally do.  Also the button in the UO merges trunk into branches.  Just my two cents.

samcoe reviewed (approved):
@rwe Thanks for making the requested changes. This looks good to ship now! 

As for merge vs rebase conversation, we do not have a merge strategy requirement for this repository so whichever you prefer is okay with us.
